
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="A tool kit to manage many variant on desktop computer">
      
      
      
        <link rel="canonical" href="https://natir.github.io/variantplaner/usages/">
      
      
        <link rel="prev" href="../license/">
      
      
        <link rel="next" href="../reference/variantplaner/">
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.4.3, mkdocs-material-9.1.16">
    
    
      
        <title>Usages - VariantPlaner</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.26e3688c.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.ecc896b0.min.css">
      
      

    
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_markdown_exec_ansi.css">
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../css/material.css">
    
      <link rel="stylesheet" href="../css/mkdocstrings.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="lime">
  
    
    
      <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#variantplaner" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="VariantPlaner" class="md-header__button md-logo" aria-label="VariantPlaner" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            VariantPlaner
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usages
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="lime"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
            
              <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_2" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
              </label>
            
          
            
            
            
            <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_2">
            
              <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
              </label>
            
          
        </form>
      
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/natir/variantplaner" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    natir/variantplaner
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        Home
      </a>
    </li>
  

      
        
  
  
    
  


  <li class="md-tabs__item">
    <a href="./" class="md-tabs__link md-tabs__link--active">
      Usages
    </a>
  </li>

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../reference/variantplaner/" class="md-tabs__link">
        Code Reference
      </a>
    </li>
  

      
        
  
  


  
  
  
    <li class="md-tabs__item">
      <a href="../contributing/" class="md-tabs__link">
        Development
      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="VariantPlaner" class="md-nav__button md-logo" aria-label="VariantPlaner" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    VariantPlaner
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/natir/variantplaner" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    natir/variantplaner
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
      
      
      
        <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
          Home
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_1">
          <span class="md-nav__icon md-icon"></span>
          Home
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Overview
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        Changelog
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../credits/" class="md-nav__link">
        Credits
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../license/" class="md-nav__link">
        License
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Usages
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Usages
      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-planner-presentation" class="md-nav__link">
    Variant planner presentation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#vcf2parquet" class="md-nav__link">
    vcf2parquet
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structuration-of-data" class="md-nav__link">
    Structuration of data
  </a>
  
    <nav class="md-nav" aria-label="Structuration of data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#merge-all-variant" class="md-nav__link">
    Merge all variant
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genotypes-structuration" class="md-nav__link">
    Genotypes structuration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#by-samples" class="md-nav__link">
    By samples
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#by-variants" class="md-nav__link">
    By variants
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#add-annotations" class="md-nav__link">
    Add annotations
  </a>
  
    <nav class="md-nav" aria-label="Add annotations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#snpeff-annotations" class="md-nav__link">
    Snpeff annotations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clinvar-annotations" class="md-nav__link">
    Clinvar annotations
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#querying" class="md-nav__link">
    Querying
  </a>
  
    <nav class="md-nav" aria-label="Querying">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#polars-cli" class="md-nav__link">
    polars-cli
  </a>
  
    <nav class="md-nav" aria-label="polars-cli">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#count-variants" class="md-nav__link">
    Count variants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-variants-from-annotations" class="md-nav__link">
    Filter variants from annotations:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-sample-have-variant" class="md-nav__link">
    Get sample have variant
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#duckdb" class="md-nav__link">
    duckdb
  </a>
  
    <nav class="md-nav" aria-label="duckdb">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#count-variants_1" class="md-nav__link">
    Count variants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-variants-from-annotations_1" class="md-nav__link">
    Filter variants from annotations:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-sample-have-variant_1" class="md-nav__link">
    Get sample have variant
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
      
      
      
        <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
          Code Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Code Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
      
      
      
        
        
        <div class="md-nav__link md-nav__link--index ">
          <a href="../reference/variantplaner/">variantplaner</a>
          
            <label for="__nav_3_1">
              <span class="md-nav__icon md-icon"></span>
            </label>
          
        </div>
      
      <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3_1">
          <span class="md-nav__icon md-icon"></span>
          variantplaner
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/cli/" class="md-nav__link">
        cli
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/exception/" class="md-nav__link">
        exception
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/extract/" class="md-nav__link">
        extract
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/generate/" class="md-nav__link">
        generate
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_5" >
      
      
      
        
        
        <div class="md-nav__link md-nav__link--index ">
          <a href="../reference/variantplaner/io/">io</a>
          
            <label for="__nav_3_1_5">
              <span class="md-nav__icon md-icon"></span>
            </label>
          
        </div>
      
      <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_1_5_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3_1_5">
          <span class="md-nav__icon md-icon"></span>
          io
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/io/csv/" class="md-nav__link">
        csv
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/io/ped/" class="md-nav__link">
        ped
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/io/vcf/" class="md-nav__link">
        vcf
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/normalization/" class="md-nav__link">
        normalization
      </a>
    </li>
  

            
          
            
              
  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1_7" >
      
      
      
        
        
        <div class="md-nav__link md-nav__link--index ">
          <a href="../reference/variantplaner/struct/">struct</a>
          
            <label for="__nav_3_1_7">
              <span class="md-nav__icon md-icon"></span>
            </label>
          
        </div>
      
      <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_3_1_7_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_3_1_7">
          <span class="md-nav__icon md-icon"></span>
          struct
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/struct/genotypes/" class="md-nav__link">
        genotypes
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reference/variantplaner/struct/variants/" class="md-nav__link">
        variants
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
      
      
      <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
      
      
      
        <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
          Development
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Development
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../contributing/" class="md-nav__link">
        Contributing
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../code_of_conduct/" class="md-nav__link">
        Code of Conduct
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../coverage/" class="md-nav__link">
        Coverage report
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../benchmark/" class="md-nav__link">
        Benchmark
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#setup" class="md-nav__link">
    Setup
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#download-data" class="md-nav__link">
    Download data
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#variant-planner-presentation" class="md-nav__link">
    Variant planner presentation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#vcf2parquet" class="md-nav__link">
    vcf2parquet
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#structuration-of-data" class="md-nav__link">
    Structuration of data
  </a>
  
    <nav class="md-nav" aria-label="Structuration of data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#merge-all-variant" class="md-nav__link">
    Merge all variant
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genotypes-structuration" class="md-nav__link">
    Genotypes structuration
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#by-samples" class="md-nav__link">
    By samples
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#by-variants" class="md-nav__link">
    By variants
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#add-annotations" class="md-nav__link">
    Add annotations
  </a>
  
    <nav class="md-nav" aria-label="Add annotations">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#snpeff-annotations" class="md-nav__link">
    Snpeff annotations
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clinvar-annotations" class="md-nav__link">
    Clinvar annotations
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#querying" class="md-nav__link">
    Querying
  </a>
  
    <nav class="md-nav" aria-label="Querying">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#polars-cli" class="md-nav__link">
    polars-cli
  </a>
  
    <nav class="md-nav" aria-label="polars-cli">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#count-variants" class="md-nav__link">
    Count variants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-variants-from-annotations" class="md-nav__link">
    Filter variants from annotations:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-sample-have-variant" class="md-nav__link">
    Get sample have variant
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#duckdb" class="md-nav__link">
    duckdb
  </a>
  
    <nav class="md-nav" aria-label="duckdb">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#count-variants_1" class="md-nav__link">
    Count variants
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter-variants-from-annotations_1" class="md-nav__link">
    Filter variants from annotations:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get-sample-have-variant_1" class="md-nav__link">
    Get sample have variant
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="variantplaner">VariantPlaner<a class="headerlink" href="#variantplaner" title="Permanent link">&para;</a></h1>
<p>variantplaner is a set of tools that converts a large set of vcf's into an interoperable data structure in an efficient way.</p>
<p>To show the capabilities of the variant planner we will use a small example.</p>
<h2 id="setup">Setup<a class="headerlink" href="#setup" title="Permanent link">&para;</a></h2>
<p>This tutorial assume you are on unix like system, you have python setup and you <a href="https://github.com/natir/variantplaner#installation">install variantplaner</a></p>
<p>Requirements list:</p>
<ul>
<li>curl</li>
<li>gunzip</li>
</ul>
<p>Optional:</p>
<ul>
<li><a href="https://github.com/manojkarthick/pqrs">pqrs</a></li>
<li>gnu-parallel</li>
</ul>
<p>Quering dataset:</p>
<ul>
<li><a href="https://crates.io/crates/polars-cli">polars-cli</a></li>
<li><a href="https://duckdb.org/">duckdb</a></li>
</ul>
<h2 id="download-data">Download data<a class="headerlink" href="#download-data" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>-p<span class="w"> </span>vp_tuto/vcf/
<span class="nb">cd</span><span class="w"> </span>vp_tuto
<span class="nv">URI_ROOT</span><span class="o">=</span><span class="s2">&quot;https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release&quot;</span>
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/NA12878_HG001/latest/GRCh38/HG001_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG001.vcf
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/AshkenazimTrio/HG002_NA24385_son/latest/GRCh38/HG002_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG002.vcf
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/AshkenazimTrio/HG003_NA24149_father/latest/GRCh38/HG003_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG003.vcf
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/AshkenazimTrio/HG004_NA24143_mother/latest/GRCh38/HG004_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG004.vcf
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/ChineseTrio/HG006_NA24694_father/latest/GRCh38/HG006_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG006.vcf
curl<span class="w"> </span><span class="si">${</span><span class="nv">URI_ROOT</span><span class="si">}</span>/ChineseTrio/HG007_NA24695_mother/latest/GRCh38/HG007_GRCh38_1_22_v4.2.1_benchmark.vcf.gz<span class="w"> </span><span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>vcf/HG007.vcf
</code></pre></div>
<h2 id="variant-planner-presentation">Variant planner presentation<a class="headerlink" href="#variant-planner-presentation" title="Permanent link">&para;</a></h2>
<p>variantplaner is a python module with command line tools, it is composed of several subcommands (they will be detailed later) but has two global options, one for parallelization and another for the level of verbosity you want.
<div class="highlight"><pre><span></span><code>Usage: variantplaner [OPTIONS] COMMAND [ARGS]...

  Run VariantPlanner.

Options:
  -t, --threads INTEGER RANGE  Number of threads usable  [default: 1; x&gt;=0]
  -v, --verbose                Verbosity level  [0&lt;=x&lt;=4]
  --help                       Show this message and exit.

Commands:
  annotations  Convert an annotation variation file in parquet.
  metadata     Convert an metadata file in parquet.
  parquet2vcf  Convert parquet in vcf.
  struct       Struct operation on parquet file.
  vcf2parquet  Convert a vcf in parquet.
</code></pre></div></p>
<h2 id="vcf2parquet">vcf2parquet<a class="headerlink" href="#vcf2parquet" title="Permanent link">&para;</a></h2>
<p>First step is convert vcf data in <a href="https://en.wikipedia.org/wiki/Apache_Parquet">parquet</a> it's a column oriented format with better performance.</p>
<p>We split vcf in two part on for variant information and another for genotype information.</p>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>-p<span class="w"> </span>variants<span class="w"> </span>genotypes/samples/
</code></pre></div>
<div class="highlight"><pre><span></span><code>for vcf_path in $(ls vcf/*.vcf)
do
vcf_basename=$(basename ${vcf_path} .vcf)
variantplaner -t 4 vcf2parquet -i ${vcf_path} \
-v variants/${vcf_basename}.parquet \
-g genotypes/samples/${vcf_basename}.parquet \
-f GT:PS:DP:ADALL:AD:GQ
done
</code></pre></div>
<details>
<summary>gnu-parallel method</summary>
<div class="highlight"><pre><span></span><code>find<span class="w"> </span>vcf<span class="w"> </span>-type<span class="w"> </span>f<span class="w"> </span>-name<span class="w"> </span>*.vcf<span class="w"> </span>-exec<span class="w"> </span>basename<span class="w"> </span><span class="o">{}</span><span class="w"> </span>.vcf<span class="w"> </span><span class="se">\;</span><span class="w"> </span><span class="p">|</span><span class="w"> </span><span class="se">\</span>
parallel<span class="w"> </span>variantplaner<span class="w"> </span>-t<span class="w"> </span><span class="m">2</span><span class="w"> </span>vcf2parquet<span class="w"> </span>-i<span class="w"> </span>vcf/<span class="o">{}</span>.vcf<span class="w"> </span>-v<span class="w"> </span>variants/<span class="o">{}</span>.parquet<span class="w"> </span><span class="se">\</span>
-g<span class="w"> </span>genotypes/samples/<span class="o">{}</span>.parquet<span class="w"> </span>-f<span class="w"> </span>GT:PS:DP:ADALL:AD:GQ
</code></pre></div>
</details>
<p>Parquet variants file contains 5 column:</p>
<ul>
<li>chr: Chromosome name, X -&gt; 22, Y -&gt; 23, MT -&gt; 24</li>
<li>pos: Position of variant</li>
<li>ref: Reference sequence</li>
<li>alt: Alternative sequence</li>
<li>id: An hash of other value colision isn't check but highly improbable <a href="/variantplaner/reference/variantplaner/normalization/#variantplaner.normalization.add_variant_id">check api documentation</a></li>
</ul>
<details>
<summary>variants parquet file content</summary>
<p>You can inspect content of parquet file generate with pqrs
<div class="highlight"><pre><span></span><code>pqrs<span class="w"> </span>head<span class="w"> </span>variants/HG001.parquet
<span class="o">{</span>id:<span class="w"> </span><span class="m">17886044532216650390</span>,<span class="w"> </span>chr:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>pos:<span class="w"> </span><span class="m">783006</span>,<span class="w"> </span>ref:<span class="w"> </span><span class="s2">&quot;A&quot;</span>,<span class="w"> </span>alt:<span class="w"> </span><span class="s2">&quot;G&quot;</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">7513336577790240873</span>,<span class="w"> </span>chr:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>pos:<span class="w"> </span><span class="m">783175</span>,<span class="w"> </span>ref:<span class="w"> </span><span class="s2">&quot;T&quot;</span>,<span class="w"> </span>alt:<span class="w"> </span><span class="s2">&quot;C&quot;</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">17987040642944149052</span>,<span class="w"> </span>chr:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>pos:<span class="w"> </span><span class="m">784860</span>,<span class="w"> </span>ref:<span class="w"> </span><span class="s2">&quot;T&quot;</span>,<span class="w"> </span>alt:<span class="w"> </span><span class="s2">&quot;C&quot;</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">10342734968077036194</span>,<span class="w"> </span>chr:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>pos:<span class="w"> </span><span class="m">785417</span>,<span class="w"> </span>ref:<span class="w"> </span><span class="s2">&quot;G&quot;</span>,<span class="w"> </span>alt:<span class="w"> </span><span class="s2">&quot;A&quot;</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">890514037559296207</span>,<span class="w"> </span>chr:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>pos:<span class="w"> </span><span class="m">797392</span>,<span class="w"> </span>ref:<span class="w"> </span><span class="s2">&quot;G&quot;</span>,<span class="w"> </span>alt:<span class="w"> </span><span class="s2">&quot;A&quot;</span><span class="o">}</span>
</code></pre></div></p>
</details>
<p>Parquet genotypes file contains column:</p>
<ul>
<li>id: Same as variant id</li>
<li>gt: vcf GT value 1 -&gt; heterozygote 2 -&gt; homozygote (phasing information is lost)</li>
<li>ps: Phase set in which this variant falls</li>
<li>dp: vcf DP coverage of the variant for this sample</li>
<li>adall: Net allele depths across all datasets</li>
<li>ad: vcf AD per allele reads depth</li>
<li>gq: vcf GQ quality of variant for this sample</li>
</ul>
<details>
<summary>genotypes parquet file content</summary>
<p>You can inspect content of parquet file generate with pqrs
<div class="highlight"><pre><span></span><code>pqrs<span class="w"> </span>head<span class="w"> </span>genotypes/HG001.parquet
<span class="o">{</span>id:<span class="w"> </span><span class="m">17886044532216650390</span>,<span class="w"> </span>sample:<span class="w"> </span><span class="s2">&quot;HG001&quot;</span>,<span class="w"> </span>gt:<span class="w"> </span><span class="m">2</span>,<span class="w"> </span>ps:<span class="w"> </span>null,<span class="w"> </span>dp:<span class="w"> </span><span class="m">652</span>,<span class="w"> </span>adall:<span class="w"> </span><span class="o">[</span><span class="m">16</span>,<span class="w"> </span><span class="m">234</span><span class="o">]</span>,<span class="w"> </span>ad:<span class="w"> </span><span class="o">[</span><span class="m">0</span>,<span class="w"> </span><span class="m">82</span><span class="o">]</span>,<span class="w"> </span>gq:<span class="w"> </span><span class="m">312</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">7513336577790240873</span>,<span class="w"> </span>sample:<span class="w"> </span><span class="s2">&quot;HG001&quot;</span>,<span class="w"> </span>gt:<span class="w"> </span><span class="m">2</span>,<span class="w"> </span>ps:<span class="w"> </span>null,<span class="w"> </span>dp:<span class="w"> </span><span class="m">639</span>,<span class="w"> </span>adall:<span class="w"> </span><span class="o">[</span><span class="m">0</span>,<span class="w"> </span><span class="m">218</span><span class="o">]</span>,<span class="w"> </span>ad:<span class="w"> </span><span class="o">[</span><span class="m">0</span>,<span class="w"> </span><span class="m">84</span><span class="o">]</span>,<span class="w"> </span>gq:<span class="w"> </span><span class="m">194</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">17987040642944149052</span>,<span class="w"> </span>sample:<span class="w"> </span><span class="s2">&quot;HG001&quot;</span>,<span class="w"> </span>gt:<span class="w"> </span><span class="m">2</span>,<span class="w"> </span>ps:<span class="w"> </span>null,<span class="w"> </span>dp:<span class="w"> </span><span class="m">901</span>,<span class="w"> </span>adall:<span class="w"> </span><span class="o">[</span><span class="m">105</span>,<span class="w"> </span><span class="m">406</span><span class="o">]</span>,<span class="w"> </span>ad:<span class="w"> </span><span class="o">[</span><span class="m">0</span>,<span class="w"> </span><span class="m">74</span><span class="o">]</span>,<span class="w"> </span>gq:<span class="w"> </span><span class="m">301</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">10342734968077036194</span>,<span class="w"> </span>sample:<span class="w"> </span><span class="s2">&quot;HG001&quot;</span>,<span class="w"> </span>gt:<span class="w"> </span><span class="m">2</span>,<span class="w"> </span>ps:<span class="w"> </span>null,<span class="w"> </span>dp:<span class="w"> </span><span class="m">820</span>,<span class="w"> </span>adall:<span class="w"> </span><span class="o">[</span><span class="m">125</span>,<span class="w"> </span><span class="m">383</span><span class="o">]</span>,<span class="w"> </span>ad:<span class="w"> </span><span class="o">[</span><span class="m">0</span>,<span class="w"> </span><span class="m">70</span><span class="o">]</span>,<span class="w"> </span>gq:<span class="w"> </span><span class="m">339</span><span class="o">}</span>
<span class="o">{</span>id:<span class="w"> </span><span class="m">890514037559296207</span>,<span class="w"> </span>sample:<span class="w"> </span><span class="s2">&quot;HG001&quot;</span>,<span class="w"> </span>gt:<span class="w"> </span><span class="m">1</span>,<span class="w"> </span>ps:<span class="w"> </span>null,<span class="w"> </span>dp:<span class="w"> </span><span class="m">760</span>,<span class="w"> </span>adall:<span class="w"> </span><span class="o">[</span><span class="m">161</span>,<span class="w"> </span><span class="m">142</span><span class="o">]</span>,<span class="w"> </span>ad:<span class="w"> </span><span class="o">[</span><span class="m">25</span>,<span class="w"> </span><span class="m">37</span><span class="o">]</span>,<span class="w"> </span>gq:<span class="w"> </span><span class="m">147</span><span class="o">}</span>
</code></pre></div></p>
</details>
<h2 id="structuration-of-data">Structuration of data<a class="headerlink" href="#structuration-of-data" title="Permanent link">&para;</a></h2>
<h3 id="merge-all-variant">Merge all variant<a class="headerlink" href="#merge-all-variant" title="Permanent link">&para;</a></h3>
<p>We can now aggregate all variant present in our dataset to perform this operation we use divide to conquer merge methode by generate temporary file. By default file are write in <code>/tmp</code> but you can control where these files are write by set <code>TMPDIR</code>, <code>TEMP</code> or <code>TMP</code> directory.</p>
<div class="highlight"><pre><span></span><code><span class="nv">input</span><span class="o">=</span><span class="k">$(</span>ls<span class="w"> </span>variants/<span class="w"> </span><span class="p">|</span><span class="w"> </span>xargs<span class="w"> </span>-I<span class="w"> </span><span class="o">{}</span><span class="w"> </span>-x<span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;-i variants/&quot;</span><span class="o">{}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"> </span><span class="s1">&#39; &#39;</span><span class="k">)</span>
variantplaner<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>struct<span class="w"> </span><span class="k">$(</span><span class="nb">echo</span><span class="w"> </span><span class="nv">$input</span><span class="k">)</span><span class="w"> </span>variants<span class="w"> </span>-o<span class="w"> </span>variants.parquet
</code></pre></div>
<p>File <code>variants.parquet</code> containt all uniq variants present in dataset.</p>
<h3 id="genotypes-structuration">Genotypes structuration<a class="headerlink" href="#genotypes-structuration" title="Permanent link">&para;</a></h3>
<h3 id="by-samples">By samples<a class="headerlink" href="#by-samples" title="Permanent link">&para;</a></h3>
<p>This structurations data is already down in vcf2parquet setp check content of <code>genotypes/samples</code>:
<div class="highlight"><pre><span></span><code>âžœ<span class="w"> </span>ls<span class="w"> </span>genotypes/samples
HG001.parquet<span class="w">  </span>HG002.parquet<span class="w">  </span>HG003.parquet<span class="w">  </span>HG004.parquet<span class="w">  </span>HG006.parquet<span class="w">  </span>HG007.parquet
</code></pre></div></p>
<h3 id="by-variants">By variants<a class="headerlink" href="#by-variants" title="Permanent link">&para;</a></h3>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>-p<span class="w"> </span>genotypes/variants/
<span class="nv">input</span><span class="o">=</span><span class="k">$(</span>ls<span class="w"> </span>genotypes/samples/<span class="w"> </span><span class="p">|</span><span class="w"> </span>xargs<span class="w"> </span>-I<span class="w"> </span><span class="o">{}</span><span class="w"> </span>-x<span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;-i genotypes/samples/&quot;</span><span class="o">{}</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>tr<span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"> </span><span class="s1">&#39; &#39;</span><span class="k">)</span>
variantplaner<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>struct<span class="w"> </span><span class="k">$(</span><span class="nb">echo</span><span class="w"> </span><span class="nv">$input</span><span class="k">)</span><span class="w"> </span>genotypes<span class="w"> </span>-p<span class="w"> </span>genotypes/variants/
</code></pre></div>
<p>All genotypes information are split in <a href="https://duckdb.org/docs/data/partitioning/hive_partitioning">hive like structure</a> to optimize request on data.</p>
<h2 id="add-annotations">Add annotations<a class="headerlink" href="#add-annotations" title="Permanent link">&para;</a></h2>
<p>To work on your variant you probably need and annotations.</p>
<h3 id="snpeff-annotations">Snpeff annotations<a class="headerlink" href="#snpeff-annotations" title="Permanent link">&para;</a></h3>
<p>First convert your unique variants in parquet format (<code>variants.parquet</code>) in vcf:
<div class="highlight"><pre><span></span><code>variantplaner<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>parquet2vcf<span class="w"> </span>-i<span class="w"> </span>variants.parquet<span class="w"> </span>-o<span class="w"> </span>variants.vcf
</code></pre></div></p>
<p><code>parquet2vcf</code> subcommand have many more option but we didn't need it now.</p>
<p>Next annotate this <code>variants.vcf</code> <a href="https://pcingola.github.io/SnpEff/">with snpeff</a>, we assume you generate a file call <code>variants.snpeff.vcf</code>.</p>
<p>To convert annotated vcf in parquet, keep 'ANN' info column and rename vcf id column in snpeff_id you can run:
<div class="highlight"><pre><span></span><code>variantplaner<span class="w"> </span>-t<span class="w"> </span><span class="m">8</span><span class="w"> </span>annotations<span class="w"> </span>-i<span class="w"> </span>variants.snpeff.vcf<span class="w"> </span>-o<span class="w"> </span>snpeff.parquet<span class="w"> </span>vcf<span class="w"> </span>-i<span class="w"> </span>ANN<span class="w"> </span>-r<span class="w"> </span>snpeff_id
</code></pre></div></p>
<p>If you didn't set any value of option <code>-i</code> in vcf subsubcommand all info column are keep.</p>
<h3 id="clinvar-annotations">Clinvar annotations<a class="headerlink" href="#clinvar-annotations" title="Permanent link">&para;</a></h3>
<p>Download last clinvar version:</p>
<div class="highlight"><pre><span></span><code>mkdir<span class="w"> </span>annotations
curl<span class="w"> </span>https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/clinvar.vcf.gz<span class="w"> </span><span class="se">\</span>
<span class="p">|</span><span class="w"> </span>gunzip<span class="w"> </span>-<span class="w"> </span>&gt;<span class="w"> </span>annotations/clinvar.vcf
</code></pre></div>
<p>Convert clinvar vcf file in parquet file:</p>
<div class="highlight"><pre><span></span><code>variantplaner<span class="w"> </span>annotations<span class="w"> </span>-i<span class="w"> </span>annotations/clinvar.vcf<span class="w"> </span>-o<span class="w"> </span>annotations/clinvar.parquet<span class="w"> </span>vcf<span class="w"> </span>-r<span class="w"> </span>clinvar_id
</code></pre></div>
<p>Parquet file produce contains many columns:</p>
<ul>
<li>clinvar_id: content of vcf id column if option <code>-r</code> is not set column name is <code>vid</code></li>
<li>id: variantplaner id</li>
<li>All INFO filed</li>
</ul>
<p>Annotations subcommand try to make match between vcf info type and parquet type.</p>
<details>
<summary><code>annotations/clinvar.parquet</code> file content</summary>
<div class="highlight"><pre><span></span><code>pqrs<span class="w"> </span>head<span class="w"> </span>annotations/clinvar.parquet
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2205837</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">11650605831284591550</span>,<span class="w"> </span>AF_ESP:<span class="w"> </span>null,<span class="w"> </span>AF_EXAC:<span class="w"> </span>null,<span class="w"> </span>AF_TGP:<span class="w"> </span>null,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2193183</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDNINCL:<span class="w"> </span>null,<span class="w"> </span>CLNDISDB:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;MeSH:D030342&quot;</span>,<span class="w"> </span><span class="s2">&quot;MedGen:C0950123&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDISDBINCL:<span class="w"> </span>null,<span class="w"> </span>CLNHGVS:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;NC_000001.11:g.69134A&gt;G&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNREVSTAT:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;criteria_provided&quot;</span>,<span class="w"> </span><span class="s2">&quot;_single_submitter&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIG:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Likely_benign&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIGCONF:<span class="w"> </span>null,<span class="w"> </span>CLNSIGINCL:<span class="w"> </span>null,<span class="w"> </span>CLNVC:<span class="w"> </span><span class="s2">&quot;single_nucleotide_variant&quot;</span>,<span class="w"> </span>CLNVCSO:<span class="w"> </span><span class="s2">&quot;SO:0001483&quot;</span>,<span class="w"> </span>CLNVI:<span class="w"> </span>null,<span class="w"> </span>DBVARID:<span class="w"> </span>null,<span class="w"> </span>GENEINFO:<span class="w"> </span><span class="s2">&quot;OR4F5:79501&quot;</span>,<span class="w"> </span>MC:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;SO:0001583|missense_variant&quot;</span><span class="o">]</span>,<span class="w"> </span>ORIGIN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;1&quot;</span><span class="o">]</span>,<span class="w"> </span>RS:<span class="w"> </span>null<span class="o">}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2252161</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">2295086632353399847</span>,<span class="w"> </span>AF_ESP:<span class="w"> </span>null,<span class="w"> </span>AF_EXAC:<span class="w"> </span>null,<span class="w"> </span>AF_TGP:<span class="w"> </span>null,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2238986</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDNINCL:<span class="w"> </span>null,<span class="w"> </span>CLNDISDB:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;MeSH:D030342&quot;</span>,<span class="w"> </span><span class="s2">&quot;MedGen:C0950123&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDISDBINCL:<span class="w"> </span>null,<span class="w"> </span>CLNHGVS:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;NC_000001.11:g.69581C&gt;G&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNREVSTAT:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;criteria_provided&quot;</span>,<span class="w"> </span><span class="s2">&quot;_single_submitter&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIG:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Uncertain_significance&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIGCONF:<span class="w"> </span>null,<span class="w"> </span>CLNSIGINCL:<span class="w"> </span>null,<span class="w"> </span>CLNVC:<span class="w"> </span><span class="s2">&quot;single_nucleotide_variant&quot;</span>,<span class="w"> </span>CLNVCSO:<span class="w"> </span><span class="s2">&quot;SO:0001483&quot;</span>,<span class="w"> </span>CLNVI:<span class="w"> </span>null,<span class="w"> </span>DBVARID:<span class="w"> </span>null,<span class="w"> </span>GENEINFO:<span class="w"> </span><span class="s2">&quot;OR4F5:79501&quot;</span>,<span class="w"> </span>MC:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;SO:0001583|missense_variant&quot;</span><span class="o">]</span>,<span class="w"> </span>ORIGIN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;1&quot;</span><span class="o">]</span>,<span class="w"> </span>RS:<span class="w"> </span>null<span class="o">}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2396347</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">11033100074712141168</span>,<span class="w"> </span>AF_ESP:<span class="w"> </span>null,<span class="w"> </span>AF_EXAC:<span class="w"> </span>null,<span class="w"> </span>AF_TGP:<span class="w"> </span>null,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2386655</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDNINCL:<span class="w"> </span>null,<span class="w"> </span>CLNDISDB:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;MeSH:D030342&quot;</span>,<span class="w"> </span><span class="s2">&quot;MedGen:C0950123&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDISDBINCL:<span class="w"> </span>null,<span class="w"> </span>CLNHGVS:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;NC_000001.11:g.69682G&gt;A&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNREVSTAT:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;criteria_provided&quot;</span>,<span class="w"> </span><span class="s2">&quot;_single_submitter&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIG:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Uncertain_significance&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIGCONF:<span class="w"> </span>null,<span class="w"> </span>CLNSIGINCL:<span class="w"> </span>null,<span class="w"> </span>CLNVC:<span class="w"> </span><span class="s2">&quot;single_nucleotide_variant&quot;</span>,<span class="w"> </span>CLNVCSO:<span class="w"> </span><span class="s2">&quot;SO:0001483&quot;</span>,<span class="w"> </span>CLNVI:<span class="w"> </span>null,<span class="w"> </span>DBVARID:<span class="w"> </span>null,<span class="w"> </span>GENEINFO:<span class="w"> </span><span class="s2">&quot;OR4F5:79501&quot;</span>,<span class="w"> </span>MC:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;SO:0001583|missense_variant&quot;</span><span class="o">]</span>,<span class="w"> </span>ORIGIN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;1&quot;</span><span class="o">]</span>,<span class="w"> </span>RS:<span class="w"> </span>null<span class="o">}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2288999</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">10487392163259126218</span>,<span class="w"> </span>AF_ESP:<span class="w"> </span>null,<span class="w"> </span>AF_EXAC:<span class="w"> </span>null,<span class="w"> </span>AF_TGP:<span class="w"> </span>null,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2278803</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDNINCL:<span class="w"> </span>null,<span class="w"> </span>CLNDISDB:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;MeSH:D030342&quot;</span>,<span class="w"> </span><span class="s2">&quot;MedGen:C0950123&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDISDBINCL:<span class="w"> </span>null,<span class="w"> </span>CLNHGVS:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;NC_000001.11:g.69769T&gt;C&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNREVSTAT:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;criteria_provided&quot;</span>,<span class="w"> </span><span class="s2">&quot;_single_submitter&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIG:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Uncertain_significance&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIGCONF:<span class="w"> </span>null,<span class="w"> </span>CLNSIGINCL:<span class="w"> </span>null,<span class="w"> </span>CLNVC:<span class="w"> </span><span class="s2">&quot;single_nucleotide_variant&quot;</span>,<span class="w"> </span>CLNVCSO:<span class="w"> </span><span class="s2">&quot;SO:0001483&quot;</span>,<span class="w"> </span>CLNVI:<span class="w"> </span>null,<span class="w"> </span>DBVARID:<span class="w"> </span>null,<span class="w"> </span>GENEINFO:<span class="w"> </span><span class="s2">&quot;OR4F5:79501&quot;</span>,<span class="w"> </span>MC:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;SO:0001583|missense_variant&quot;</span><span class="o">]</span>,<span class="w"> </span>ORIGIN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;1&quot;</span><span class="o">]</span>,<span class="w"> </span>RS:<span class="w"> </span>null<span class="o">}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2351346</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">5356120651941363990</span>,<span class="w"> </span>AF_ESP:<span class="w"> </span>null,<span class="w"> </span>AF_EXAC:<span class="w"> </span>null,<span class="w"> </span>AF_TGP:<span class="w"> </span>null,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2333177</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDNINCL:<span class="w"> </span>null,<span class="w"> </span>CLNDISDB:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;MeSH:D030342&quot;</span>,<span class="w"> </span><span class="s2">&quot;MedGen:C0950123&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNDISDBINCL:<span class="w"> </span>null,<span class="w"> </span>CLNHGVS:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;NC_000001.11:g.69995G&gt;C&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNREVSTAT:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;criteria_provided&quot;</span>,<span class="w"> </span><span class="s2">&quot;_single_submitter&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIG:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Uncertain_significance&quot;</span><span class="o">]</span>,<span class="w"> </span>CLNSIGCONF:<span class="w"> </span>null,<span class="w"> </span>CLNSIGINCL:<span class="w"> </span>null,<span class="w"> </span>CLNVC:<span class="w"> </span><span class="s2">&quot;single_nucleotide_variant&quot;</span>,<span class="w"> </span>CLNVCSO:<span class="w"> </span><span class="s2">&quot;SO:0001483&quot;</span>,<span class="w"> </span>CLNVI:<span class="w"> </span>null,<span class="w"> </span>DBVARID:<span class="w"> </span>null,<span class="w"> </span>GENEINFO:<span class="w"> </span><span class="s2">&quot;OR4F5:79501&quot;</span>,<span class="w"> </span>MC:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;SO:0001583|missense_variant&quot;</span><span class="o">]</span>,<span class="w"> </span>ORIGIN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;1&quot;</span><span class="o">]</span>,<span class="w"> </span>RS:<span class="w"> </span>null<span class="o">}</span>
</code></pre></div>
</details>
<p>With option of subcommand vcf <code>-i</code> you can select which column are include in parquet file
For example command:
<div class="highlight"><pre><span></span><code>variantplaner<span class="w"> </span>annotations<span class="w"> </span>-i<span class="w"> </span>annotations/clinvar.vcf<span class="w"> </span>-o<span class="w"> </span>annotations/clinvar.parquet<span class="w"> </span>vcf<span class="w"> </span>-r<span class="w"> </span>clinvar_id<span class="w"> </span>-i<span class="w"> </span>ALLELEID<span class="w"> </span>-i<span class="w"> </span>CLNDN<span class="w"> </span>-i<span class="w"> </span>AF_ESP<span class="w"> </span>-i<span class="w"> </span>GENEINFO
</code></pre></div></p>
<p>Produce a <code>annotations/clinvar.parquet</code> with columns:</p>
<ul>
<li>clinvar_id</li>
<li>id</li>
<li>ALLELEID</li>
<li>CLNDN</li>
</ul>
<details>
<summary><code>annotations/clinvar.parquet</code> file content</summary>
<div class="highlight"><pre><span></span><code>âžœ<span class="w"> </span>pqrs<span class="w"> </span>head<span class="w"> </span>annotations/clinvar.parquet
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2205837</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">11650605831284591550</span>,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2193183</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2252161</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">2295086632353399847</span>,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2238986</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2396347</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">11033100074712141168</span>,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2386655</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2288999</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">10487392163259126218</span>,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2278803</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]}</span>
<span class="o">{</span>clinvar_id:<span class="w"> </span><span class="m">2351346</span>,<span class="w"> </span>id:<span class="w"> </span><span class="m">5356120651941363990</span>,<span class="w"> </span>ALLELEID:<span class="w"> </span><span class="m">2333177</span>,<span class="w"> </span>CLNDN:<span class="w"> </span><span class="o">[</span><span class="s2">&quot;Inborn_genetic_diseases&quot;</span><span class="o">]}</span>
</code></pre></div>
</details>
<h2 id="querying">Querying<a class="headerlink" href="#querying" title="Permanent link">&para;</a></h2>
<p>You could use <a href="https://crates.io/crates/polars-cli">polars-cli</a> or <a href="https://duckdb.org/">duckdb</a> to interogate your variants database.</p>
<h3 id="polars-cli">polars-cli<a class="headerlink" href="#polars-cli" title="Permanent link">&para;</a></h3>
<h4 id="count-variants">Count variants<a class="headerlink" href="#count-variants" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="err">ã€‰</span><span class="k">select</span><span class="w"> </span><span class="k">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">);</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w"> </span><span class="k">count</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="c1">---     â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">u32</span><span class="w">     </span><span class="err">â”‚</span>
<span class="err">â•žâ•â•â•â•â•â•â•â•â•â•¡</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">7852699</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>
<details>
<summary>check result with <code>pqrs</code></summary>
<p>We can check we have same result with pqrs</p>
<div class="highlight"><pre><span></span><code>âžœ<span class="w"> </span>pqrs<span class="w"> </span>rowcount<span class="w"> </span>variants.parquet
File<span class="w"> </span>Name:<span class="w"> </span>variants.parquet:<span class="w"> </span><span class="m">7852699</span><span class="w"> </span>rows
</code></pre></div>
</details>
<h4 id="filter-variants-from-annotations">Filter variants from annotations:<a class="headerlink" href="#filter-variants-from-annotations" title="Permanent link">&para;</a></h4>
<p>Get all variant with a AF_ESP upper than 0.9999</p>
<div class="highlight"><pre><span></span><code><span class="err">ã€‰</span><span class="k">select</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="k">ref</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="p">,</span><span class="w"> </span><span class="n">AF_ESP</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;annotations/clinvar.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="k">c</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="k">c</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">where</span><span class="w"> </span><span class="n">AF_ESP</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">.</span><span class="mi">9999</span><span class="p">;</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">chr</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">pos</span><span class="w">      </span><span class="err">â”†</span><span class="w"> </span><span class="k">ref</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">alt</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">AF_ESP</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="c1">--- â”† ---      â”† --- â”† --- â”† ---     â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">u8</span><span class="w">  </span><span class="err">â”†</span><span class="w"> </span><span class="n">u64</span><span class="w">      </span><span class="err">â”†</span><span class="w"> </span><span class="n">str</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">str</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">f64</span><span class="w">     </span><span class="err">â”‚</span>
<span class="err">â•žâ•â•â•â•â•â•ªâ•â•â•â•â•â•â•â•â•â•â•ªâ•â•â•â•â•â•ªâ•â•â•â•â•â•ªâ•â•â•â•â•â•â•â•â•â•¡</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">10</span><span class="w">  </span><span class="err">â”†</span><span class="w"> </span><span class="mi">16901372</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="k">G</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">0</span><span class="p">.</span><span class="mi">99992</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">11</span><span class="w">  </span><span class="err">â”†</span><span class="w"> </span><span class="mi">78121030</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">A</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">0</span><span class="p">.</span><span class="mi">99992</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>
<h4 id="get-sample-have-variant">Get sample have variant<a class="headerlink" href="#get-sample-have-variant" title="Permanent link">&para;</a></h4>
<p>Get all variant and sample with GENEINFO equal to 'SAMD11:148398'</p>
<div class="highlight"><pre><span></span><code><span class="err">ã€‰</span><span class="k">select</span><span class="w"> </span><span class="k">distinct</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="k">ref</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;genotypes/samples/*&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="k">g</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="k">g</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;annotations/clinvar.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="n">a</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">WHERE</span><span class="w"> </span><span class="n">GENEINFO</span><span class="o">=</span><span class="s1">&#39;SAMD11:148398&#39;</span><span class="p">;</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">chr</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">pos</span><span class="w">    </span><span class="err">â”†</span><span class="w"> </span><span class="k">ref</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">alt</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">sample</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="c1">--- â”† ---    â”† --- â”† --- â”† ---    â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">u8</span><span class="w">  </span><span class="err">â”†</span><span class="w"> </span><span class="n">u64</span><span class="w">    </span><span class="err">â”†</span><span class="w"> </span><span class="n">str</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">str</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">str</span><span class="w">    </span><span class="err">â”‚</span>
<span class="err">â•žâ•â•â•â•â•â•ªâ•â•â•â•â•â•â•â•â•ªâ•â•â•â•â•â•ªâ•â•â•â•â•â•ªâ•â•â•â•â•â•â•â•â•¡</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG003</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG001</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG007</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG004</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="err">â€¦</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="err">â€¦</span><span class="w">      </span><span class="err">â”†</span><span class="w"> </span><span class="err">â€¦</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="err">â€¦</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="err">â€¦</span><span class="w">      </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942934</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="k">G</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG003</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">943937</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG004</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG002</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">1</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”†</span><span class="w"> </span><span class="n">T</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="k">C</span><span class="w">   </span><span class="err">â”†</span><span class="w"> </span><span class="n">HG006</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>
<h3 id="duckdb">duckdb<a class="headerlink" href="#duckdb" title="Permanent link">&para;</a></h3>
<h4 id="count-variants_1">Count variants<a class="headerlink" href="#count-variants_1" title="Permanent link">&para;</a></h4>
<div class="highlight"><pre><span></span><code><span class="n">D</span><span class="w"> </span><span class="k">select</span><span class="w"> </span><span class="k">count</span><span class="p">(</span><span class="o">*</span><span class="p">)</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">);</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">count_star</span><span class="p">()</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">    </span><span class="n">int64</span><span class="w">     </span><span class="err">â”‚</span>
<span class="err">â”œâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤</span>
<span class="err">â”‚</span><span class="w">      </span><span class="mi">7852699</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>
<details>
<summary>check result with <code>pqrs</code></summary>
<p>We can check we have same result with pqrs</p>
<div class="highlight"><pre><span></span><code>âžœ<span class="w"> </span>pqrs<span class="w"> </span>rowcount<span class="w"> </span>variants.parquet
File<span class="w"> </span>Name:<span class="w"> </span>variants.parquet:<span class="w"> </span><span class="m">7852699</span><span class="w"> </span>rows
</code></pre></div>
</details>
<h4 id="filter-variants-from-annotations_1">Filter variants from annotations:<a class="headerlink" href="#filter-variants-from-annotations_1" title="Permanent link">&para;</a></h4>
<p>Get all variant with a AF_ESP upper than 0.9999</p>
<div class="highlight"><pre><span></span><code><span class="n">D</span><span class="w"> </span><span class="k">select</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="k">ref</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="p">,</span><span class="w"> </span><span class="n">AF_ESP</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;annotations/clinvar.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="k">c</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="k">c</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">where</span><span class="w"> </span><span class="n">AF_ESP</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">.</span><span class="mi">9999</span><span class="p">;</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w">  </span><span class="n">chr</span><span class="w">  </span><span class="err">â”‚</span><span class="w">   </span><span class="n">pos</span><span class="w">    </span><span class="err">â”‚</span><span class="w">   </span><span class="k">ref</span><span class="w">   </span><span class="err">â”‚</span><span class="w">   </span><span class="n">alt</span><span class="w">   </span><span class="err">â”‚</span><span class="w"> </span><span class="n">AF_ESP</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">uint8</span><span class="w"> </span><span class="err">â”‚</span><span class="w">  </span><span class="n">uint64</span><span class="w">  </span><span class="err">â”‚</span><span class="w"> </span><span class="nb">varchar</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="nb">varchar</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">double</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”œâ”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤</span>
<span class="err">â”‚</span><span class="w">    </span><span class="mi">10</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">16901372</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="k">G</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">0</span><span class="p">.</span><span class="mi">99992</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">    </span><span class="mi">11</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">78121030</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">A</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">0</span><span class="p">.</span><span class="mi">99992</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>
<h4 id="get-sample-have-variant_1">Get sample have variant<a class="headerlink" href="#get-sample-have-variant_1" title="Permanent link">&para;</a></h4>
<p>Get all variant and sample with GENEINFO equal to 'SAMD11:148398'</p>
<div class="highlight"><pre><span></span><code><span class="n">D</span><span class="w"> </span><span class="k">select</span><span class="w"> </span><span class="k">distinct</span><span class="w"> </span><span class="n">chr</span><span class="p">,</span><span class="w"> </span><span class="n">pos</span><span class="p">,</span><span class="w"> </span><span class="k">ref</span><span class="p">,</span><span class="w"> </span><span class="n">alt</span><span class="p">,</span><span class="w"> </span><span class="n">sample</span><span class="w"> </span><span class="k">from</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;variants.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">v</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;genotypes/samples/*&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="k">g</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="k">g</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">left</span><span class="w"> </span><span class="k">join</span><span class="w"> </span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;annotations/clinvar.parquet&#39;</span><span class="p">)</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="k">on</span><span class="w"> </span><span class="n">v</span><span class="p">.</span><span class="n">id</span><span class="o">=</span><span class="n">a</span><span class="p">.</span><span class="n">id</span><span class="w"> </span><span class="k">WHERE</span><span class="w"> </span><span class="n">GENEINFO</span><span class="o">=</span><span class="s1">&#39;SAMD11:148398&#39;</span><span class="p">;</span>
<span class="err">â”Œâ”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€â”€â”€â”</span>
<span class="err">â”‚</span><span class="w">  </span><span class="n">chr</span><span class="w">  </span><span class="err">â”‚</span><span class="w">  </span><span class="n">pos</span><span class="w">   </span><span class="err">â”‚</span><span class="w">   </span><span class="k">ref</span><span class="w">   </span><span class="err">â”‚</span><span class="w">   </span><span class="n">alt</span><span class="w">   </span><span class="err">â”‚</span><span class="w"> </span><span class="n">sample</span><span class="w">  </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w"> </span><span class="n">uint8</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">uint64</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="nb">varchar</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="nb">varchar</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="nb">varchar</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â”œâ”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¼â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG002</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942934</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="k">G</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG002</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG003</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942934</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="k">G</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG003</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG007</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">943937</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG007</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG001</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG004</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">943937</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG004</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”‚</span><span class="w">     </span><span class="mi">1</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="mi">942451</span><span class="w"> </span><span class="err">â”‚</span><span class="w"> </span><span class="n">T</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="k">C</span><span class="w">       </span><span class="err">â”‚</span><span class="w"> </span><span class="n">HG006</span><span class="w">   </span><span class="err">â”‚</span>
<span class="err">â”œâ”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”´â”€â”€â”€â”€â”€â”€â”€â”€â”€â”¤</span>
<span class="err">â”‚</span><span class="w"> </span><span class="mi">10</span><span class="w"> </span><span class="k">rows</span><span class="w">                            </span><span class="mi">5</span><span class="w"> </span><span class="n">columns</span><span class="w"> </span><span class="err">â”‚</span>
<span class="err">â””â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”˜</span>
</code></pre></div>





                
              </article>
            </div>
          
          
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
            Back to top
          </button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/natir" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.top", "search.highlight", "search.suggest"], "search": "../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.a51614de.min.js"></script>
      
    
  </body>
</html>